### PR TITLE
WINC-490: Remove downloading CNI binaries

### PR DESCRIPTION
**Pull request by PratikMahajan**
This PR builds the CNI plugins and then transfers them to the Windows
VM instead of downloading the CNI plugins directly from github.

Also, changes the way we copy files to the Windows VM
Now, instead of taking individual files to be transferred,
we will transfer the folders to the Windows VM according to preset source
and destination similar to what we're doing in WMCO. Nested folders wont
be copied over.

### CODE REVIEW

File: internal/test/wmcb/wmcb_test.go [hunk 1]
```diff
@@ -58,9 +58,9 @@ var (
 		Value:  "Windows",
 		Effect: v1.TaintEffectNoSchedule,
 	}
-	// filesToBeTransferred holds the list of files that needs to be transferred to the Windows VM
-	filesToBeTransferred = flag.String("filesToBeTransferred", "",
-		"Comma separated list of files to be transferred")
+	// foldersToBeTransferred holds the list of folders that needs to be transferred to the Windows VM
+	foldersToBeTransferred = flag.String("foldersToBeTransferred", "",
```
Comment (LorbusChris) [inline internal/test/wmcb/wmcb_test.go:59]: `directories` or `dirs` would be better imo

File: internal/test/windows/windows.go [hunk 2]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
 	return nil
 }
 
+// CopyDirectory copies the files from the directory on the local host to the remote Windows VM
+// This does not copy nested directories
+func (w *Windows) CopyDirectory(localDir string, remoteDir string) error {
```
Comment (LorbusChris) [inline internal/test/windows/windows.go:92]: nit : maybe a better name would be `TransferDirectory` for this func, to be in sync with the flag

File: internal/test/windows/windows.go [hunk 3]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
...
+
+	sftp, err := sftp.NewClient(w.SSHClient)
+	if err != nil {
+		return fmt.Errorf("sftp initialization failed: %v", err)
+	}
+	defer sftp.Close()
+
+	log.Printf("Copying %v directory to Windows : %v", localDir, remoteDir)
+
+	// creating a local directory to store the files and directories from remote directory.
```
Comment (PratikMahajan) [inline internal/test/windows/windows.go:105]: change comment

**Comment by LorbusChris** [comment]
A lot of vocabulary here. I think we should agree on one of:
- `Directories`, `Dirs` or `Folders`
- `copy`, `transfer` or `move`

File: Dockerfile.tools [hunk 4]
```diff
@@ -82,6 +92,6 @@ COPY --from=build /build/wmcb_unit_test.exe .
 COPY --from=build /build/wmcb_e2e_test.exe .
 COPY --from=build /build/internal/test/wmcb/test-wmcb .
 
-ENV PATH="${PATH}:/home/test"
+ENV PATH="${PATH}:/payload"
```
Comment (sebsoto) [inline Dockerfile.tools:95]: Why does this need to be added to the PATH?

File: internal/test/windows/windows.go [hunk 5]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
...
+		return fmt.Errorf("cannot retrieve remote directory without a ssh client")
+	}
+
+	sftp, err := sftp.NewClient(w.SSHClient)
+	if err != nil {
+		return fmt.Errorf("sftp initialization failed: %v", err)
+	}
+	defer sftp.Close()
+
+	log.Printf("Copying %v directory to Windows : %v", localDir, remoteDir)
```
Comment (sebsoto) [inline internal/test/windows/windows.go:103]: nit: %s instead of %v

File: internal/test/windows/windows.go [hunk 6]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
 	return nil
 }
 
+// CopyDirectory copies the files from the directory on the local host to the remote Windows VM
+// This does not copy nested directories
+func (w *Windows) CopyDirectory(localDir string, remoteDir string) error {
+	if w.SSHClient == nil {
+		return fmt.Errorf("cannot retrieve remote directory without a ssh client")
```
Comment (sebsoto) [inline internal/test/windows/windows.go:94]: Comment needs to be changed

File: internal/test/windows/windows.go [hunk 7]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
...
+	if err != nil {
+		return fmt.Errorf("sftp initialization failed: %v", err)
+	}
+	defer sftp.Close()
+
+	log.Printf("Copying %v directory to Windows : %v", localDir, remoteDir)
+
+	// creating a local directory to store the files and directories from remote directory.
+	err = sftp.MkdirAll(remoteDir)
+	if err != nil {
```
Comment (sebsoto) [inline internal/test/windows/windows.go:107]: can make this into 1 line

File: internal/test/windows/windows.go [hunk 8]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
...
+	// Get the list of all files in the directory
+	localFiles, err := ioutil.ReadDir(localDir)
+	if err != nil {
+		return fmt.Errorf("error opening local directory %s: %v", localDir, err)
+	}
+
+	for _, localFile := range localFiles {
+		localPath := filepath.Join(localDir, localFile.Name())
+		// transfer only files and not directories
+		if !localFile.IsDir() {
```
Comment (sebsoto) [inline internal/test/windows/windows.go:120]: nit: it could be preferred to do
```
if localFIle.IsDir() {
  continue
}
```
in order to reduce indentation

File: internal/test/windows/windows.go [hunk 9]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
...
+			if err = w.CopyFile(localPath, remoteDir); err != nil {
+				log.Printf("error while copying %s file to Windows : %v", localPath, err)
```
Comment (sebsoto) [inline internal/test/windows/windows.go:122]: This should return an error, as that file could be expected later

File: internal/test/wmcb/wmcb_test.go [hunk 10]
```diff
@@ -256,17 +238,10 @@ func (vm *wmcbVM) initializeTestConfigureCNIFiles(ovnHostSubnet string) error {
...
 
-	cniPkgUrl := framework.pkgs[cniPluginPkgName].getUrl()
-	cniUrl, err := url.Parse(cniPkgUrl)
-	if err != nil {
-		return fmt.Errorf("error parsing %s: %v", cniPkgUrl, err)
-	}
-
-	// Download and extract the CNI binaries on the Windows VM
-	err = vm.remoteDownloadExtract(framework.pkgs[cniPluginPkgName], remoteDir+path.Base(cniUrl.Path), winCNIDir)
+	out, err := vm.Run(fmt.Sprintf("Move-Item -Path %s -Destination %s", remoteDir+path.Base("cni"), winCNIDir), true)
```
Comment (sebsoto) [inline internal/test/wmcb/wmcb_test.go:241]: Why do this instead of just copying the files to where we want in the first place?

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: internal/test/wmcb/wmcb_test.go [hunk 11]
```diff
@@ -112,10 +90,14 @@ func TestWMCB(t *testing.T) {
 	for _, vm := range framework.WinVMs {
 		log.Printf("Testing VM: %s", vm.GetCredentials().InstanceId())
 		wVM := &wmcbVM{vm}
-		files := strings.Split(*filesToBeTransferred, ",")
-		for _, file := range files {
-			err := wVM.CopyFile(file, remoteDir)
-			require.NoError(t, err, "error copying %s to the Windows VM", file)
+		folders := strings.Split(*foldersToBeTransferred, ",")
```
Comment (VaishnaviHire) [inline internal/test/wmcb/wmcb_test.go:93]: `folder` and `directory` are used interchangeably here. Should we just go with `directory` / `dir`?

**Review by VaishnaviHire — COMMENTED** [review]
Thanks @PratikMahajan for working on this. PTAL at my comment

File: internal/test/windows/windows.go (hunk 2 above)
Comment (aravindhp) [reply internal/test/windows/windows.go:92]: But this would put it in out of sync with `CopyFile`. I suggest leaving it as `CopyDirectory`.

File: internal/test/windows/windows.go [hunk 12]
```diff
@@ -81,6 +87,45 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
 	return nil
 }
 
+// CopyDirectory copies the files from the directory on the local host to the remote Windows VM
+// This does not copy nested directories
```
Comment (aravindhp) [inline internal/test/windows/windows.go:91]: Not needed here if you are describing it in the interface.

File: internal/test/windows/windows.go (hunk 5 above)
Comment (aravindhp) [reply internal/test/windows/windows.go:103]: `Windows VM:`

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: internal/test/wmcb/wmcb_test.go (hunk 10 above)
Comment (PratikMahajan) [reply internal/test/wmcb/wmcb_test.go]: We dont know the destination while transferring the files to Windows VM. So Transferring it to a temp directory and then copying it to required directory.

File: internal/test/wmcb/wmcb_test.go [hunk 13]
```diff
@@ -58,9 +55,9 @@ var (
 		Value:  "Windows",
 		Effect: v1.TaintEffectNoSchedule,
 	}
-	// filesToBeTransferred holds the list of files that needs to be transferred to the Windows VM
-	filesToBeTransferred = flag.String("filesToBeTransferred", "",
-		"Comma separated list of files to be transferred")
+	// dirsToBeTransferred holds the list of folders that needs to be transferred to the Windows VM
```
Comment (PratikMahajan) [inline internal/test/wmcb/wmcb_test.go:58]: directories

File: internal/test/wmcb/wmcb_test.go [hunk 14]
```diff
@@ -112,10 +90,14 @@ func TestWMCB(t *testing.T) {
 	for _, vm := range framework.WinVMs {
 		log.Printf("Testing VM: %s", vm.GetCredentials().InstanceId())
 		wVM := &wmcbVM{vm}
-		files := strings.Split(*filesToBeTransferred, ",")
-		for _, file := range files {
-			err := wVM.CopyFile(file, remoteDir)
-			require.NoError(t, err, "error copying %s to the Windows VM", file)
+		folders := strings.Split(*dirsToBeTransferred, ",")
```
Comment (PratikMahajan) [inline internal/test/wmcb/wmcb_test.go:93]: directories

File: internal/test/windows/windows.go [hunk 15]
```diff
@@ -55,6 +59,8 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
 	}
 	defer ftp.Close()
 
+	log.Printf("Copying %s file to Windows : %v", filePath, remoteDir)
```
Comment (mansikulkarni96) [inline internal/test/windows/windows.go:62]: Windows vm*

File: internal/test/wmcb/wmcb_test.go [hunk 16]
```diff
@@ -256,17 +238,10 @@ func (vm *wmcbVM) initializeTestConfigureCNIFiles(ovnHostSubnet string) error {
...
 	}
 
-	cniPkgUrl := framework.pkgs[cniPluginPkgName].getUrl()
-	cniUrl, err := url.Parse(cniPkgUrl)
-	if err != nil {
-		return fmt.Errorf("error parsing %s: %v", cniPkgUrl, err)
-	}
-
-	// Download and extract the CNI binaries on the Windows VM
-	err = vm.remoteDownloadExtract(framework.pkgs[cniPluginPkgName], remoteDir+path.Base(cniUrl.Path), winCNIDir)
```
Comment (mansikulkarni96) [inline internal/test/wmcb/wmcb_test.go:266]: can we remove removeDownload and removeDownloadExtract methods, they don't seem to be used anywhere?

File: internal/test/wmcb/pkginfo.go [hunk 17]
```diff
@@ -52,31 +52,11 @@ type pkgInfo struct {
 // pkgInfoFactory returns PkgInfo specific to the package name
 func pkgInfoFactory(name pkgName, shaType string, baseUrl string, version string) (PkgInfo, error) {
 	switch name {
-	case cniPluginPkgName:
```
Comment (mansikulkarni96) [inline internal/test/wmcb/pkginfo.go:55]: is this file pkginfo required anymore?

File: internal/test/windows/windows.go (hunk 15 above)
Comment (mansikulkarni96) [inline internal/test/windows/windows.go:62]: Windows VM*

**Review by mansikulkarni96 — COMMENTED** [review]
Thanks for working on this @PratikMahajan PTAL ta my comments.

File: Dockerfile.tools [hunk 18]
```diff
@@ -72,14 +72,15 @@ ENV GOCACHE="/go/.cache"
...
-COPY internal/test/wmcb/templates templates
+WORKDIR /payload
 COPY internal/test/wmcb/powershell/ .
-
 COPY --from=build /build/ovn-kubernetes/go-controller/_output/go/bin/windows/hybrid-overlay-node.exe .
 COPY --from=build /build/kubernetes/_output/local/bin/windows/amd64/kubelet.exe .
 COPY --from=build /build/wmcb_unit_test.exe .
 COPY --from=build /build/wmcb_e2e_test.exe .
+
+WORKDIR /home/test
```
Comment (aravindhp) [inline Dockerfile.tools:92]: Why do we need a separate directory for the template? We should definitely move away from `/home/test/` either way.

File: internal/test/windows/windows.go [hunk 19]
```diff
@@ -29,6 +30,9 @@ type Windows struct {
 
 // WindowsVM is the interface for interacting with a Windows object created by the cloud provider
 type WindowsVM interface {
+	// CopyDirectory copies the files from the directory on the local host to the remote Windows VM
```
Comment (aravindhp) [inline internal/test/windows/windows.go:33]: `...to the directory on the remote Windows VM ...`

File: internal/test/windows/windows.go [hunk 20]
```diff
@@ -81,6 +87,43 @@ func (w *Windows) CopyFile(filePath, remoteDir string) error {
...
+	localFiles, err := ioutil.ReadDir(localDir)
+	if err != nil {
+		return fmt.Errorf("error opening local directory %s: %v", localDir, err)
+	}
+
+	for _, localFile := range localFiles {
+		localPath := filepath.Join(localDir, localFile.Name())
+		// transfer only files and not directories
+		if localFile.IsDir() {
+			continue
```
Comment (aravindhp) [inline internal/test/windows/windows.go:118]: Add a log saying we are skipping this directory,

File: internal/test/wmcb/wmcb_test.go (hunk 13 above)
Comment (aravindhp) [reply internal/test/wmcb/wmcb_test.go:58]: Not sure what you mean by saying `directories`.

Please replace `folders` with `directories`.

File: internal/test/wmcb/wmcb_test.go [hunk 21]
```diff
@@ -112,10 +112,14 @@ func TestWMCB(t *testing.T) {
 	for _, vm := range framework.WinVMs {
 		log.Printf("Testing VM: %s", vm.GetCredentials().InstanceId())
 		wVM := &wmcbVM{vm}
-		files := strings.Split(*filesToBeTransferred, ",")
-		for _, file := range files {
-			err := wVM.CopyFile(file, remoteDir)
-			require.NoError(t, err, "error copying %s to the Windows VM", file)
+		folders := strings.Split(*dirsToBeTransferred, ",")
+		// all the files in `/payload` directory will be transferred to `C:\Temp`
+		// for files not in `/payload`, new directory will be created in `C:\Temp` which resembles the linux folder path
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go:95]: I don't understand this comment. For one, this comment will need to be updated every time the input list of directories changes. So you need to say something to the effect of `not in the list of directories`. Moreover what do you mean by `for files not in /payload`?

>		// for files not in `/payload`, new directory will be created in `C:\Temp` which resembles the linux folder path

Where are we doing this? I think confusion is because you are using values instead of the referencing the variables here.

File: internal/test/wmcb/wmcb_test.go [hunk 22]
```diff
@@ -112,10 +112,14 @@ func TestWMCB(t *testing.T) {
...
+		for _, folder := range folders {
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go:96]: Please use `directories` or `dirs` instead of `folder` everywhere including comments. In comments please use `directories`.

File: Dockerfile.tools [hunk 23]
```diff
@@ -36,8 +36,12 @@ RUN git clone --branch release-4.6 --single-branch https://github.com/openshift/
 WORKDIR /build/ovn-kubernetes/go-controller/
 RUN make windows
 
-# TODO: Add building the CNI plugins to dockerfile and remove downloading binary
-# https://issues.redhat.com/browse/WINC-490
+# Build CNI plugins
+WORKDIR /build/
+RUN git clone --branch release-4.6 --single-branch https://github.com/openshift/containernetworking-plugins
```
Comment (aravindhp) [inline Dockerfile.tools:41]: I think we should clone WMCO and use the sub-module there. We need to do this for other components too but that can be a separate PR.

@LorbusChris maybe you can tackle this as part of updating sub-modules in WMCO.

File: internal/test/wmcb/pkginfo.go (hunk 17 above)
Comment (aravindhp) [reply internal/test/wmcb/pkginfo.go:55]: If we are removing the last user of `pkgInfo`, then `pkginfo.go` should be deleted using `git rm`.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @PratikMahajan. Please address my comments.

File: internal/test/wmcb/wmcb_test.go (hunk 21 above)
Comment (PratikMahajan) [reply internal/test/wmcb/wmcb_test.go]: Updated the comment to make it better to understand. LMK if its still confusing

File: internal/test/wmcb/wmcb_test.go [hunk 24]
```diff
@@ -112,10 +112,17 @@ func TestWMCB(t *testing.T) {
...
+		// all the files in `/payload` directory will be transferred to `remoteDir`
+		// for files not in `/payload`, new directory will be created in `remoteDir` which resembles the linux folder path
+		// for eg. `/payload/cni` will be stored in `remoteDir\cni` and `/foo/bar` will be stored in `remoteDir\foo\bar`
+		for _, dir := range dirs {
+			// remove `/payload` from the path and split the path
+			dirPath := strings.Split(strings.Replace(dir, "/payload", "", -1), "/")
+			// join the path with `\\` and place it in `remoteDir`
+			remotePath := remoteDir + "\\" + strings.Join(dirPath, "\\")
+			err := wVM.CopyDirectory(dir, remotePath)
+			require.NoError(t, err, "error copying %s to the Windows VM", dir)
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go:99]: I am a nack on this approach. You are assuming that the input will be `/payload`. Sure we are building the image that way but you cannot leak such assumptions into the test code.

**Review by aravindhp — CHANGES_REQUESTED** [review]

**Review by LorbusChris — APPROVED** [review]

**Review by aravindhp — APPROVED** [review]
Thanks for working on this, @PratikMahajan 

/approve
